keylime/da/record.py

```diff
@@ -0,0 +1,331 @@
+from keylime.fs_util import ch_dir
+
+logger = keylime_logging.init_logging("durable_attestation")
+
+
+class BaseRecordManagement(metaclass=abc.ABCMeta):
+    def __init__(self, service, key_tls_pub=None):
+
+        self.svc = service
+
+        self.ps_url = urlparse(config.get(self.svc, "persistent_store_url", fallback=""))
+        self.tl_url = urlparse(config.get(self.svc, "transparency_log_url", fallback=""))
+        self.tsa_url = urlparse(config.get(self.svc, "time_stamp_authority_url", fallback=""))
+        self.tsa_cert = config.get(self.svc, "time_stamp_authority_certs_path", fallback="")
+
+        self.st_imp_module = config.get(self.svc, "durable_attestation_import", fallback="")
+        self.st_imp_path = ".".join(self.st_imp_module.split(".")[0:-1])
+
+        self.rmv_a = ["ssl_context", "pending_event"]
+        self.rcd_fmt = config.get(self.svc, "persistent_store_format", fallback="json")
+        self.rcd_enc = config.get(self.svc, "persistent_store_encoding", fallback="")
+        self.rcd_sa = config.get(self.svc, "transparency_log_sign_algo", fallback="sha256")
+        self.tmp_d_cl = True
+        self.key_tls_priv = None
+        self.key_tls_pub = key_tls_pub
+        self.priv_key = None
+        logger.info('The "Durable Attestion" feature is stil experimental and might change in the future')
+        logger.debug('Persistent store module %s imported for "Durable Attestation"', self.st_imp_module)
+
+    @abc.abstractmethod
+    def record_create(
+        self,
+        agent_data: dict,
+        attestation_data: dict,
+        ima_policy_data: dict,
+        service: str = "auto",
+        signed_attributes: str = "auto",
+    ):
+        """Takes agent data, attestation data, ima policy data, serialize, optionally encodes it and writes to a persistent data store"""
+
+    def get_record_type(self, service: str) -> str:
+        """Determine which "service", (initial) registration or attestation is used)"""
+        if service != "auto":
+            self.svc = service
+
+        if self.svc == "registrar":
+            return "registration"
+
+        return "attestation"
+
+    def set_certs_path(self, override_service: bool = False):
+        """Get the paths for TLS certficates and key pair (used for agent data attribute signature, optionally) from the Keylime configurations"""
+        if override_service:
+            self.svc = override_service
+
+        if not self.key_tls_pub:
+            (_, self.key_tls_priv, _, _), _ = web_util.get_tls_options(self.svc, logger=logger)
+
+            self.key_tls_pub = self.key_tls_priv.replace("-private", "-public")
+
+            if self.tl_url:
+                with open(self.key_tls_priv, "rb") as fp:
+                    self.priv_key = crypto.rsa_import_privkey(fp.read())
+
+    def record_serialize(self, record_object: dict) -> bytes:
+        """Serialize, and optionally encodes, record"""
+
+        _serialized_record_object = record_object
```

Comment:
This only creates a shallow copy, which is not what we want.

Reply:
even worse, it is sharing the same object reference, that is why I am scared of mutable APIs

```python
>>> a={"a":1}
>>> b = a
>>> id(a) == id(b)
True
```

Reply:
Yes, I forgot that in Python is is not even a copy. Using `copy.deepcopy()` is probably the way to go here.

Reply:
I was reluctant on using `copy.deepcopy` for performance reasons, but can certainly implement it this way. The whole code is still at a point where no one has extensive real-world production experience with it, and thus I can go back and optimize if and when needed.